docs/src/constraints.md

```diff
@@ -1,5 +1,514 @@
+value of the objective given an infinitesimal relaxation of the constraint. For
+example, in a linear constraint, this relaxation is equivalent to `2x <= 1 + δ`
+and `2x >= 1 - δ`. If the constraint is an equality constraint, it depends on
+which direction is binding.
+
+Notably, this definition is independent of the objective sense. That is, the
+dual associated with a constraint is the same when the objective is a
+maximization and when it is a minimization. **This is different to traditional
+duality in linear programming.**
+
+The dual value associated with a constraint can be accessed using the
+[`JuMP.dual`](@ref) function. For example:
+
+```jldoctest
+julia> model = Model()
+A JuMP Model
+
+julia> @variable(model, x)
+x
+
+julia> @constraint(model, con, x <= 1)
+con : x <= 1.0
+```
+
+```@meta
+DocTestSetup = quote
+    using JuMP
+    const MOI = JuMP.MathOptInterface
+    model = Model(
+        with_optimizer(
+            MOI.Utilities.MockOptimizer,
+            JuMP.JuMPMOIModel{Float64}(),
+            eval_objective_value = false,
+            eval_variable_constraint_dual = false));
+    @variable(model, x);
+    @constraint(model, con, x <= 1);
+    @objective(model, Max, -2x);
+    JuMP.optimize!(model);
+    mock = JuMP.caching_optimizer(model).optimizer;
+    MOI.set(mock, MOI.ConstraintDual(), JuMP.optimizer_index(con), -2.0)
+end
+```
+
+```jldoctest con_duality
+julia> @objective(model, Min, -2x)
+-2 x
+
+julia> JuMP.optimize!(model)
+
+julia> JuMP.dual(con)
+-2.0
+
+julia> @objective(model, Max, 2x)
+2 x
+
+julia> JuMP.optimize!(model)
+
+julia> JuMP.dual(con)
+-2.0
+```
+
+To help users who may be less familiar with conic duality, JuMP provides the
+[`JuMP.shadow_price`](@ref) function which returns a value that can be
+interpreted as the improvement in the objective given a 1-unit change in the
+right-hand side of the constraint. [`JuMP.shadow_price`](@ref) can only be used
+on linear constraints with a `<=`, `>=`, or `==` comparison operator.
+
+In the example above, `JuMP.dual(con)` returned `-2.0` regardless of the
+optimization sense. However, in the second case when the optimization sense is
+`Max`, [`JuMP.shadow_price`](@ref) returns:
+```jldoctest con_duality
+julia> JuMP.shadow_price(con)
+2.0
+```
+
+To query the dual variables associated a variable bounds, first obtain a
+constraint reference using one of [`JuMP.UpperBoundRef`](@ref),
+[`JuMP.LowerBoundRef`](@ref), or [`JuMP.FixRef`](@ref), and then call
+[`JuMP.dual`](@ref) on the returned constraint reference.
+
+```@meta
+DocTestSetup = quote
+    using JuMP
+    const MOI = JuMP.MathOptInterface
+end
+```
+
+## Constraint containers
+
+So far, we've adding constraints one-by-one. However, just like
+[Variable containers](@ref), JuMP provides a mechanism for building groups of
+constraints in *containers*. Three types of constraint containers are supported:
+`Array`s, `JuMPArray`s, and `Dict`ionaries. We explain each of these in the
```

Comment:
The emphasis seems a bit reversed. The point of this syntax is to let you add groups of constraints compactly, and as a side effect we return containers of references that let you index into the constraints. The constraints themselves aren't in containers; the references are.